### ISSUE TITLE
`.chezmoiignore` doesn't distinguish between `!*` and `!*/`

### ISSUE THREAD

**Issue by ja1den** [twpayne/chezmoi#2507]
## Description

`.chezmoiignore` doesn't distinguish between `!*` and `!*/`.
The former should exclude all files, while the latter should only exclude directories.
Currently, both are treated as excluding all files.

This comes from the [SO link](https://stackoverflow.com/q/987142/13234077) in #261, which links to [this answer](https://stackoverflow.com/a/8025106/13234077) that explains it more clearly.

## Steps to Reproduce

Create a file `.chezmoignore` with the following content:

```
*
!*/
```

Running `chezmoi managed` should return no results (i.e., be empty).
Currently, it will return all possible results (i.e., the `*` and the `!*/` lines cancel out).

## Additional context

I need this to work in order to set up opt-in dotfile management, rather than opt-out.

The ideal workflow would be as follows:

By default, no files will be managed, preventing unnecessary configuration from being copied / templated.

To opt-in to managing a file, add a corresponding line to the `.chezmoignore`.

```
*
!*/

!.bashrc
!.config/starship.toml
```

Running `chezmoi managed` should return `.bashrc` and `.config/starship.toml` (assuming both exist).

**Comment by twpayne** [comment]
The syntax of `.chezmoiignore` files is different to `.gitignore`. From memory, the reason for this is that chezmoi needs finer grain control than `.gitignore` provides. Specifically, with `.chezmoiignore` you can ignore a directory but not its subdirectory. StackOverflow answers about `.gitignore` do not apply to `.chezmoiignore`.

chezmoi already has opt-in dotfile management: run `chezmoi add` on the files you want chezmoi to manage, all other files are ignored. As a helper, the `chezmoi managed` and `chezmoi unmanaged` commands list managed and unmanaged files and directories respectively. You do not need a `.chezmoiignore` file for opt-in dotfile management.

**Comment by ja1den** [comment]
> That said, wouldn't the example above work if it was
> ```
> **
>
> !.bashrc
> !.config/starship.toml
> ```
> instead?

This does exactly what I need. Thank you.

**Comment by bradenhilton** [comment]
@ja1den While that might be true right now, I don't think it's entirely correct.

@twpayne Does this look as expected to you?

```console
❯ tree $(chezmoi source-path) /F
.
│  .chezmoi.toml.tmpl
│  .chezmoiignore
├──.chezmoitemplates
│  └──powershell
│        profile.ps1
├──Documents
│  ├──PowerShell
│  │     Microsoft.PowerShell_profile.ps1.tmpl
│  └──WindowsPowerShell
│        Microsoft.PowerShell_profile.ps1.tmpl
└──private_dot_config
   ├──powershell
   │     Microsoft.PowerShell_profile.ps1.tmpl
   └──scoop
         modify_config.json.ps1.tmpl

❯ Get-Content "$(chezmoi source-path)/.chezmoiignore"
**

{{ if eq .chezmoi.os "windows" }}
!Documents/**
!.config/scoop/**
{{ end }}

❯ chezmoi managed
Documents
Documents/PowerShell
Documents/PowerShell/Microsoft.PowerShell_profile.ps1
Documents/WindowsPowerShell
Documents/WindowsPowerShell/Microsoft.PowerShell_profile.ps1

❯ chezmoi ignored
.config
```

It also doesn't work with `!.config/scoop/config.json`.

**Comment by ja1den** [comment]
@bradenhilton 

Does the following work for you?

```
**

{{ if eq .chezmoi.os "windows" }}
!Documents/**
!.config
!.config/scoop
!.config/scoop/**
{{ end }}
```

**Comment by ja1den** [comment]
I've noticed that while `**` successfully ignores everything correctly, it breaks `chezmoi unmanaged`.

Here's what I have at the moment.

`.chezmoiignore`:
```
**

# Bash
{{- if lookPath "bash" }}
!.bashrc
!.bash_profile
{{- end }}

# Paru
{{- if lookPath "paru" }}
!.config
!.config/paru
{{- end }}
# ...
```

My system has `bash` installed, but not `paru`.

Here's what I'm seeing:

```sh
> chezmoi managed
.bash_profile
.bashrc

> chezmoi unmanaged

> chezmoi ignored
.config
.config/paru
```

From what I can tell, `chezmoi unmanaged` should show something like the following:

```sh
> chezmoi unmanaged
.bash_history
.bash_logout
.cache
.lesshst
Desktop
Documents
Downloads
```

**Comment by ja1den** [comment]
Thinking about this more, I believe this could be resolved by making it so `.chezmoiignore` only applies to files in the source state.

Thus, there would be three possibilities for how a file is managed:
1. The file is in the source state and is not in the `.chezmoiignore` (would be printed by `chezmoi managed`).
2. The file is in the source state and is in the `.chezmoiignore` (would be printed by `chezmoi ignored`).
3. The file is not in the source state (would be printed by `chezmoi unmanaged`).

This would also fix the fact that `chezmoi add`ing a file that is not yet in the source state, but is already ignored (either explicitly or through a glob, such as `**`), does nothing.

**Comment by halostatue** [comment]
I don’t think that will work for my usage of `.chezmoiignore`, but I agree with you that there’s some confusion on how ignoring works—and how it should work—and that’s problematic.

I’m going to shift this from an issue to a discussion thread.
